src/txn/txn.c

```diff
@@ -838,111 +838,94 @@ __txn_locate_hs_record(WT_SESSION_IMPL *session, WT_CURSOR *hs_cursor, WT_PAGE *
 }
 
 /*
- * __txn_commit_timestamps_usage_check --
- *     Print warning messages when encountering unexpected timestamp usage.
+ * __txn_timestamp_usage_check --
+ *     Check if a commit will violate timestamp rules.
  */
-static inline int
-__txn_commit_timestamps_usage_check(WT_SESSION_IMPL *session, WT_TXN_OP *op, WT_UPDATE *upd)
+static inline void
+__txn_timestamp_usage_check(WT_SESSION_IMPL *session, WT_TXN_OP *op, WT_UPDATE *upd)
 {
+    WT_BTREE *btree;
     WT_TXN *txn;
     wt_timestamp_t op_ts, prev_op_durable_ts;
-    uint32_t ts_flags;
+    uint32_t flags;
     char ts_string[2][WT_TS_INT_STRING_SIZE];
+    const char *name;
     bool txn_has_ts;
 
-    /*
-     * Do not check for timestamp usage in recovery as it is possible that timestamps may be out of
-     * order due to WiredTiger log replay in recovery doesn't use any timestamps.
-     */
-    if (F_ISSET(S2C(session), WT_CONN_RECOVERING))
-        return (0);
-
+    btree = op->btree;
     txn = session->txn;
+    flags = btree->dhandle->ts_flags;
+    name = btree->dhandle->name;
     txn_has_ts = F_ISSET(txn, WT_TXN_HAS_TS_COMMIT | WT_TXN_HAS_TS_DURABLE);
 
-#define WT_COMMIT_TS_VERB_PREFIX "Commit timestamp unexpected usage: "
+    /* Skip timestamp usage checks unless both assert and usage configurations are set. */
+    if (!LF_ISSET(WT_DHANDLE_TS_ASSERT_WRITE))
+        return;
+    if (!LF_ISSET(WT_DHANDLE_TS_ALWAYS | WT_DHANDLE_TS_MIXED_MODE | WT_DHANDLE_TS_NEVER |
+          WT_DHANDLE_TS_ORDERED))
+        return;
 
-    /* If this transaction did not touch any table configured for verbose logging, we're done. */
-    if (!F_ISSET(txn, WT_TXN_VERB_TS_WRITE))
-        return (0);
+    /* Timestamps are ignored on logged files. */
+    if (!F_ISSET(S2C(session), WT_CONN_IN_MEMORY) && !F_ISSET(btree, WT_BTREE_NO_LOGGING))
```

Comment:
I struggle to read basic predicates, but I believe this means in-memory always checks timestamp usage, regardless of whether the table is logged?

Based on our recent findings with in-memory I presume you'll want to remove the in memory check here? We're just digging a hole where one day (maybe today?) inmemory will start failing because it enforces more constraints than durable WT.

Reply:
Yes, exactly; this is one of the places that will change based on WT-8449. Hopefully this one goes in first, but regardless, the merge won't be too bad.